### PR TITLE
Add 'std.format.read.formattedRead' overloads to return a Tuple with values read

### PR DESCRIPTION
**Pull request by iK4tsu**
This enhancement was mentioned in a Discord discussion as a potentially better alternative to the existing `std.file.slurp`. Although `std.file.slurp` does a similar thing, it is restricted to reading from files. This feature could be extended to any input range and replace the currently implemented in `std.file` since that one "combines multiple things together that would be more useful if they were separate and composable" (quoted from @pbackus).

### CODE REVIEW

**Comment by iK4tsu** [comment]
It seems the documentation is not style-compliant. How do I document the template and the function correctly? I tried to look for existing examples but couldn't find any.

**Comment by pbackus** [comment]
You need to document both the outer template and the inner function.

**Comment by iK4tsu** [comment]
update: autosquashed commits since CI passed

**Review by ljmf00 — COMMENTED** [review]
Interesting approach. I'm basically in favor of this. I think it's worth adding a changelog entry for this. Can you also please add a unittest for a CTFE `formattedRead`?

File: std/format/read.d [hunk 1]
```diff
@@ -663,6 +664,86 @@ if (isSomeString!(typeof(fmt)))
...
+
+/// The format string can be checked at compile-time:
+@safe pure unittest
+{
+    import std.typecons : tuple;
+
+    auto expected = tuple("hello", 124, 34.5);
+    auto result = "hello!124:34.5".formattedRead!("%s!%s:%s", string, int, double);
+    assert(result == expected);
+}
```
Comment (iK4tsu) [inline std/format/read.d:745]:
```suggestion
/// The format string can be checked at compile-time:
@safe pure unittest
{
    import std.typecons : tuple;

    enum expected = tuple("hello", 124, 34.5);
    enum result = "hello!124:34.5".formattedRead!("%s!%s:%s", string, int, double);
    static assert(result == expected);
}
```

Comment (iK4tsu) [reply std/format/read.d:745]: > Can you also please add a unittest for a CTFE formattedRead?

Would this be enough?

File: std/format/read.d [hunk 2]
```diff
@@ -663,6 +664,86 @@ if (isSomeString!(typeof(fmt)))
...
+    }
+}
+
+/// The format string can be checked at compile-time:
+@safe pure unittest
+{
+    import std.typecons : tuple;
+
+    auto expected = tuple("hello", 124, 34.5);
+    auto result = "hello!124:34.5".formattedRead!("%s!%s:%s", string, int, double);
```
Comment (atilaneves) [inline std/format/read.d:743]: If the format string can be checked at compile-time, why would one need to pass in types?

**Comment by pbackus** [comment]
Just leaving the tuple members default-initialized when there's nothing to read is not good enough, because it gives the user no way to distinguish between a default value read from the input and a default value returned due to incorrectly formatted input. For example:

```d
auto result1 = formattedRead!int("0", "%d");
auto result2 = formattedRead!int("", "%d");
assert(result1 == result2); // no difference
```

We need some other way of signalling an incomplete read in situations like this.

File: std/format/read.d (hunk 2 above)
Comment (iK4tsu) [reply std/format/read.d:743]: The format string is not responsible for unformatting into the types the user wants, it's only to know the places in the input string that need to be unformatted and to tell the formatted the restrictions of the characters it's reading. Using `%f` can still be valid for both `float` and `double` for example, and it asserts statically when the types are not format-compatible.
```d
float value;
"123".formattedRead!"%d"(value);
// Error: static assert:  "incompatible format character for floating point argument: %d"
```

It also asserts on orphan format specifiers or if the number of arguments is higher than the required by the format string. This is useful for generic code when we don't have total control over the argument quantity or the format string. The format string might be generated during compile time causing it to assert if some malformation occurs, or the variadic template arguments might be calculated at compile time asserting for the same reasons.

Comment (iK4tsu) [reply std/format/read.d:743]: I actually think I found a bug in `checkFormatException` for `%f` specifier.
```d
int i;
"123".formattedRead!"%f"(i);
// should static assert but instead throws at run time with:
// std.format.FormatException@/dlang/dmd/linux/bin64/../../src/phobos/std/format/internal/read.d(104): Wrong unformat specifier '%f' for int
```

**Comment by iK4tsu** [comment]
We could extend the type to return the number of arguments read as the first tuple argument. But that would defeat the practicality of this, and kinda force the user to handle the value, even if it means mapping it out.

We also could flip the signature and add an optional out parameter for the number of arguments read. But for that, I think the user might as well just use the already existing overload that receives a tuple since they would have to handle the value separately anyway.
```d
Tuple!(int, float) value;
char[] line = "1 2.125".dup;
if (!formattedRead(line, "%s %s", value) == 2) ...
...
```

Or perhaps adding a sentinel template argument for that: `countReadArgs` or something like that. But I think this other way to distinguish these cases should be something that does not take away the practicality of the implementation and should not get in the way of the user/be forced upon them.

**Comment by pbackus** [comment]
Another option is to throw an exception—iirc this is what `slurp` does.

**Comment by iK4tsu** [comment]
Not really what `slurp` does. It does throw but only when the format is malformed, so a normal format exception or there are remaining characters in the current line being parsed.
This throws:
```d
writeFile("deleteme", "123 ");
scope(exit) removeFile("deleteme");
"deleteme".slurp!(int)("%d");
// object.Exception@/dlang/dmd/linux/bin64/../../src/phobos/std/file.d(5191): Trailing characters at the end of line: ` '
```

This is runs fine:
```d
writeFile("deleteme", "123");
scope(exit) removeFile("deleteme");
assert("deleteme".slurp!(int, string)("%d %s").equal([tuple(123, "")]));
```

File: std/format/read.d (hunk 2 above)
Comment (ljmf00) [reply std/format/read.d:743]: > I actually think I found a bug in `checkFormatException` for the `%f` format.
> 
> ```d
> int i;
> "123".formattedRead!"%f"(i);
> // should static assert but instead throws at run time with:
> // std.format.FormatException@/dlang/dmd/linux/bin64/../../src/phobos/std/format/internal/read.d(104): Wrong unformat specifier '%f' for int
> ```

Could you please file a bug report for this?

**Comment by ljmf00** [comment]
> Just leaving the tuple members default-initialized when there's nothing to read is not good enough, because it gives the user no way to distinguish between a default value read from the input and a default value returned due to incorrectly formatted input. For example:
> 
> ```d
> auto result1 = formattedRead!int("0", "%d");
> auto result2 = formattedRead!int("", "%d");
> assert(result1 == result2); // no difference
> ```
> 
> We need some other way of signalling an incomplete read in situations like this.

I think the way is to throw an exception, but I would say to also provide an overload to allow the user to specify a tuple to avoid throwing and allocating on these cases.

**Comment by pbackus** [comment]
@ljmf00 The existing overloads of `formattedRead` already accept a tuple as an argument; see the example here: https://github.com/dlang/phobos/blob/v2.101.1/std/format/read.d#L355-L361

**Comment by ljmf00** [comment]
> @ljmf00 The existing overloads of `formattedRead` already accept a tuple as an argument; see the example here: https://github.com/dlang/phobos/blob/v2.101.1/std/format/read.d#L355-L361

Ok cool. Then I think if the user wants to use this prettier version, my suggestion is to raise an exception. If the user wants to use it for control flow they have alternatives and exceptions are zero cost if not raised, so it's fine to do it.

**Comment by iK4tsu** [comment]
With only an exception being thrown there's no real way to handle failures. The best I could think of is:
```d
only("123 num", "123")
    .formattedRead!(int, string)("%d %s")
    .map!nullable
    .handle!(<ExceptionName>, RangePrimitive.front, (e, r) => typeof(r.front).init );
    // [Tuple!(int, string)(123, "num"), Nullable.null]
```

But if the exception also stored info on how many arguments were read and the tuple result then the user could handle it better if they wanted to. There may be a high possibility the user would like to keep the read arguments.

**Comment by ljmf00** [comment]
> Another bug I found while playing with `formattedRead`: when using `Tuple` and the templated format version, the check is not performed on the tuple arguments but on the tuple itself leading to a static assert:
> 
> ```d
> Tuple!(int, string) tup;
> "123".formattedRead!"%d %s"(tup);
> // Error: static assert:  "Orphan format specifier: %s"
> ```
> 
> The non-templated version executes just fine.

File a bug report.

**Comment by ljmf00** [comment]
@atilaneves we need to decide an acceptable design for this feature. What's your thoughts about the ideas we discussed above?

**Comment by iK4tsu** [comment]
> > Another bug I found while playing with `formattedRead`: when using `Tuple` and the templated format version, the check is not performed on the tuple arguments but on the tuple itself leading to a static assert:
> > ```d
> > Tuple!(int, string) tup;
> > "123".formattedRead!"%d %s"(tup);
> > // Error: static assert:  "Orphan format specifier: %s"
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The non-templated version executes just fine.
> 
> File a bug report.

done: #8661

**Comment by atilaneves** [comment]
I agree with @pbackus and think that throwing an exception is the right way to deal with missing data.

**Comment by ljmf00** [comment]
> I agree with @pbackus and think that throwing an exception is the right way to deal with missing data.

I think it makes sense too. So you are in favor of the overall change with the particularity of using an exception, on these cases?

@iK4tsu can you add a changelog entry so we can get this in?

**Comment by ljmf00** [comment]
Also, use an exception, as suggested by @atilaneves

**Comment by iK4tsu** [comment]
Sorry for the delay on this. Now both functions **optionally** throw a `FormatException` if the format range ends early and not all arguments are filled. This optional behavior **is activated by default** but can be ignored by passing `No.exhaustive` as an argument. Is this a good compromise, or is it best if the exception is always enforced?

**Comment by iK4tsu** [comment]
update: rebased with master

**Comment by pbackus** [comment]
What's the use-case for `No.exhaustive`? If the user wants to handle partial reads, it seems to me that they should use the overload that actually reports how much was read.

**Comment by iK4tsu** [comment]
> What's the use-case for No.exhaustive? If the user wants to handle partial reads, it seems to me that they should use the overload that actually reports how much was read.

The main usage is to comply with the current `slurp` implementation. Currently `slurp` does nothing if the range ends early and arguments are not filled. By following `slurp`'s implementation, it can be replaced by this version of `formattedRead` without breaking changes.

The other usage is when the user does not care about the number of args filled at all but wants to chain the call. With the overload, they would ignore the result but they would be forced to declare all _default initialized_ variables beforehand. With the forced exception, they would be forced to use `std.exception.handle`.

By being optional it still follows the current design. The "error" value can be ignored, and all unfilled arguments remain unchanged, in this case, with their default initializers, while providing a compatible way to replace the alternative/s.

One issue remains IMO if the exception is indeed thrown and caught. The user knows something happened with the format, but not in specific. If it failed due to unfilled arguments, the user won't know it - by the FormatException - nor how many were filled. Should I change the Exception to a more specific one that also stores how many arguments were filled? It could potentially store the tuple with all arguments as well, but that would require it to be a templated exception. Just throwing an exception doesn't really match what can be done with the overload.

**Comment by pbackus** [comment]
> The main usage is to comply with the current `slurp` implementation.

IMO the way `slurp` handles partial reads is a mistake (and arguably a bug), not something we should copy.

> The other usage is when the user does not care about the number of args filled at all but wants to chain the call.

Under what circumstances, exactly, would a user call `formattedRead` without caring whether the data they asked for is actually read?

> Should I change the Exception to a more specific one that also stores how many arguments were filled? It could potentially store the tuple with all arguments as well, but that would require it to be a templated exception.

I think the plain `FormatException` is fine. The purpose of this new overload is to make things more convenient in the common case, not to cover every possible use case. Most of the time, the outcomes you care about are "all arguments read successfully" and "something went wrong." If a user wants more detailed information, they can use the existing overloads.

**Comment by iK4tsu** [comment]
> IMO the way slurp handles partial reads is a mistake (and arguably a bug), not something we should copy.

Well, but wouldn't that make it harder to deprecate `slurp` in favor of this by Phobos' standards?

> Under what circumstances, exactly, would a user call formattedRead without caring whether the data they asked for is actually read?

The same situations where users write code like:
```d
int a, b, c;

// don't care about the output
// don't care if the format ends early
formattedRead!someFormat(a, b, c);

// use a b c
```
 or any slurp usage.

But once again, this was more to facilitate the transition from one to another and to keep the existing nonthrowing logic of the other overloads.

**Comment by pbackus** [comment]
> Well, but wouldn't that make it harder to deprecate `slurp` in favor of this by Phobos' standards?

As far as I know there is no plan to deprecate or remove `slurp`.

If you mean, would it make it harder to *refactor* `slurp` to use the new `formattedRead` overloads internally—yes, it would. But I would rather have a good API for future users of the new `formattedRead` overloads and force `slurp` to use a workaround than have a good API for `slurp` and force future users of `formattedRead` to avoid a footgun (which is what `No.exhaustive` would be).

> The same situations where users write code like:
> ```
> int a, b, c;
> 
> // don't care about the output
> // don't care if the format ends early
> formattedRead!someFormat(a, b, c);
>
> // use a b c
> ```

In other words: situations where users write buggy code because the API makes it easy to do the wrong thing. :)

Since we have the opportunity here to design a new API, I think we should try to avoid making it easy to do the wrong thing—which meas not including a "do the wrong thing" flag.

**Comment by iK4tsu** [comment]
> Since we have the opportunity here to design a new API, I think we should try to avoid making it easy to do the wrong thing—which meas not including a "do the wrong thing" flag.

Fair enough. I will change it and create a changelog entry when I have a bit of free time.

**Review by ljmf00 — APPROVED** [review]
Looks good. Fix the style issues reported by buildkite

**Comment by LightBender** [comment]
@iK4tsu if you can resolve the style issues found by buildkite and DAutoTest, we'll put the auto-merge tag on this.

**Comment by iK4tsu** [comment]
Sorry for the delay on this.

Changelog:
* removed the exhaustive flag from all overloads; now it always enforces all arguments must be read

**Comment by iK4tsu** [comment]
Changelog:
* removed unused unit test import
* added a changelog entry

File: std/format/read.d [hunk 3]
```diff
@@ -692,6 +693,99 @@ if (isSomeString!(typeof(fmt)))
...
+    */
+    Tuple!Args formattedRead(Range, Char)(auto ref Range r, const(Char)[] fmt)
+    {
+        import core.lifetime : forward;
+        import std.exception : enforce;
+        import std.format : FormatException;
+
+        Tuple!Args args;
+        const numArgsFilled = .formattedRead(forward!r, fmt, args.expand);
+        enforce!FormatException(numArgsFilled == Args.length);
```
Comment (pbackus) [inline std/format/read.d:739]: 1. `enforce!FormatException` should be replaced with the alias `enforceFmt` for consistency with other code in this package.
2. Please add a descriptive error message to the exception.

File: std/format/read.d [hunk 4]
```diff
@@ -692,6 +693,99 @@ if (isSomeString!(typeof(fmt)))
...
+    import std.typecons : Flag, Tuple, Yes;
+    Tuple!Args formattedRead(Range)(auto ref Range r)
+    {
+        import core.lifetime : forward;
+        import std.exception : enforce;
+        import std.format : FormatException;
+
+        Tuple!Args args;
+        const numArgsFilled = .formattedRead!fmt(forward!r, args.expand);
+        enforce!FormatException(numArgsFilled == Args.length);
```
Comment (pbackus) [inline std/format/read.d:770]: Same here.

**Review by pbackus — CHANGES_REQUESTED** [review]

File: std/format/read.d (hunk 3 above)
Comment (iK4tsu) [reply std/format/read.d]: done

File: std/format/read.d (hunk 4 above)
Comment (iK4tsu) [reply std/format/read.d]: done

File: std/format/read.d (hunk 3 above)
Comment (iK4tsu) [reply std/format/read.d]: Oops, done

**Comment by iK4tsu** [comment]
Changelog:
* removed unused imports from the new overload and unit tests
* updated the documentation according to the most recent change

**Review by pbackus — APPROVED** [review]
